src/fsae_drivers/fsae_can.c

```diff
@@ -0,0 +1,241 @@
+#include "fsae_can.h"
+
+#include "MY17_Can_Library.h"
+#include "error_handler.h"
+#include "board.h"
+
+static uint32_t last_bms_heartbeat_time = 0;
+static uint32_t last_bms_errors_time = 0;
+static uint32_t last_bms_cellTemps_time = 0;
+static uint32_t last_bms_packStatus_time = 0;
+
+void Receive_Vcu_Heartbeat(BMS_INPUT_T *bms_input);
+void Receive_Unknown_Message(void);
+
+void Send_Bms_Heartbeat(BMS_INPUT_T *bms_input, BMS_STATE_T *bms_state);
+void Send_Bms_Errors(void);
+void Send_Bms_CellTemps(BMS_PACK_STATUS_T * pack_status);
+void Send_Bms_PackStatus(BMS_PACK_STATUS_T * pack_status);
+
+Can_Bms_ErrorID_T bms_error_to_can_error(ERROR_T error);
+Can_Bms_ErrorID_T get_error_status(void);
+
+void Fsae_Can_Init(uint32_t baud_rate) {
+    Can_Init(baud_rate);
+}
+
+void Fsae_Can_Receive(BMS_INPUT_T *bms_input, BMS_OUTPUT_T *bms_output) {
+    UNUSED(bms_output);
+    Can_MsgID_T msgType = Can_MsgType();
+    if (msgType == Can_No_Msg) {
+      return;
+    }
+    else if (msgType == Can_Unknown_Msg) {
+      Receive_Unknown_Message();
+    }
+    else if (msgType == Can_Vcu_BmsHeartbeat_Msg){
+      Receive_Vcu_Heartbeat(bms_input);
+    }
+    else {
+        // TODO handle current messages
+    }
+}
+
+void Fsae_Can_Transmit(BMS_INPUT_T *bms_input, BMS_STATE_T *bms_state, BMS_OUTPUT_T *bms_output) {
+    UNUSED(bms_output);
+    uint32_t msTicks = bms_input->msTicks;
+    if ( (msTicks - last_bms_heartbeat_time) > BMS_HEARTBEAT_PERIOD) {
+        last_bms_heartbeat_time = msTicks;
+        Send_Bms_Heartbeat(bms_input, bms_state);
+    }
+    if ( (msTicks - last_bms_errors_time) > BMS_ERRORS_PERIOD) {
+        last_bms_errors_time = msTicks;
+        Send_Bms_Errors();
+    }
+    if ( (msTicks - last_bms_cellTemps_time) > BMS_CELL_TEMPS_PERIOD) {
+        last_bms_cellTemps_time = msTicks;
+        Send_Bms_CellTemps(bms_input->pack_status);
+    }
+    if ( (msTicks - last_bms_packStatus_time) > BMS_PACK_STATUS_PERIOD) {
+        last_bms_packStatus_time = msTicks;
+        Send_Bms_PackStatus(bms_input->pack_status);
+    }
+
+}
+
+void Receive_Vcu_Heartbeat(BMS_INPUT_T *bms_input) {
+    bms_input->last_vcu_msg_ms = bms_input->msTicks;
```

Comment:
Whoops, missed it: https://github.com/MITEVT/ltc-battery-management-system/pull/20/files#diff-803c5170888b8642f2a97e5e9423d399R146